### PR TITLE
CTL: limbs (CPU) <-> bits (logic)

### PR DESCRIPTION
**Pull request by nbgl**
Uses the linear combination `Column` API to do CTL directly between limbs in the CPU table and bits in the logic table. Removes 32 columns from the logic table.

### CODE REVIEW

File: evm/src/all_stark.rs [hunk 1]
```diff
@@ -116,11 +116,11 @@ mod tests {
...
             assert_eq!(logic::PACKED_LIMB_BITS, 16);
-            for col in logic::columns::INPUT0_PACKED {
-                row[col] = F::from_canonical_u16(rng.gen());
+            for col in logic::columns::INPUT0 {
+                row[col] = F::from_bool(rng.gen());
             }
-            for col in logic::columns::INPUT1_PACKED {
-                row[col] = F::from_canonical_u16(rng.gen());
+            for col in logic::columns::INPUT1 {
+                row[col] = F::from_bool(rng.gen());
```
Comment (nbgl) [inline evm/src/all_stark.rs:123]: Previously we sampled from 16-bit limbs (which were later decomposed to bits). Now sample the bits directly.

File: evm/src/logic.rs [hunk 2]
```diff
@@ -145,8 +127,7 @@ fn make_result<F: RichField>(lv: &mut [F; columns::NUM_COLUMNS], op: Op) {
 impl<F: RichField, const D: usize> LogicStark<F, D> {
     pub fn generate(&self, lv: &mut [F; columns::NUM_COLUMNS]) {
         let op = check_op_flags(lv);
-        check_limb_length(lv);
-        make_bit_decomposition(lv);
+        check_bits(lv);
         make_result(lv, op);
```
Comment (nbgl) [inline evm/src/logic.rs:131]: We no longer decompose the limbs into bits. Instead, we read the bits directly.

File: evm/src/logic.rs [hunk 3]
```diff
@@ -66,6 +66,7 @@ pub struct LogicStark<F, const D: usize> {
 }
 
 enum Op {
+    Zero,
```
Comment (wborgeaud) [inline evm/src/logic.rs:69]: nit: Maybe add a comment saying what this op does and why we need it.

File: evm/src/logic.rs [hunk 4]
```diff
@@ -78,62 +79,43 @@ fn check_op_flags<F: RichField>(lv: &[F; columns::NUM_COLUMNS]) -> Op {
...
-        columns::INPUT1_PACKED
+        columns::limb_bit_cols_for_input(columns::INPUT0),
+        columns::limb_bit_cols_for_input(columns::INPUT1),
     ) {
-        let limb_in0 = lv[limb_in0_col].to_canonical_u64();
-        let limb_in1 = lv[limb_in1_col].to_canonical_u64();
+        let limb_in0: u64 = limb_in0_cols
+            .enumerate()
+            .map(|(i, col)| lv[col].to_canonical_u64() << i)
+            .sum();
```
Comment (wborgeaud) [inline evm/src/logic.rs:112]: nit: It seems like this conversion `bits -> u64/usize` is done in a few places. Maybe we could add a function for it.

**Review by wborgeaud — APPROVED** [review]
Looks good to me 👍

**Comment by nbgl** [comment]
Those were both good points, @wborgeaud! I’ve made those changes. Would you mind having a look at them? They’re just big enough that I wouldn’t be comfortable merging without another review.

**Review by wborgeaud — APPROVED** [review]
Thanks, looks good 👍

**Pull request merged**